File: src/services/storage/file-meta.ts
/** Canonical on-disk shape for `.tie/files/{id}/meta.json`. Extra fields are preserved. */

export type FileKind = 'file' | 'directory'
export type FileMode = 'copy' | 'link'
export type FileLocatorType = 'desktop' | 'android' | 'display'

export interface FileLocator {
  type: FileLocatorType
  desktopPath?: string
  androidUri?: string
  displayPath: string
}

export interface FileMeta {
  id: string
  title: string
  kind: FileKind
  mode: FileMode
  sourceId?: string
  mime: string
  ext: string
  size: number
  sourcePath: string
  storedPath: string
  locator: FileLocator
  contentHash?: string | null
  sha256?: string | null
  createdAt: string
  updatedAt: string
  preview?: string | null
  entryCount?: { files?: number; dirs?: number } | null
  mtime?: string | null
}

function text(value: unknown) {
  return String(value ?? '').trim()
}

function looksLikeContentUri(value: string) {
  return /^content:\/\//i.test(value)
}

function locatorFrom(raw: Record<string, unknown>, sourcePath: string, storedPath: string): FileLocator {
  const existing = raw.locator && typeof raw.locator === 'object' ? raw.locator as Record<string, unknown> : null
  const androidUri = text(existing?.androidUri)
    || (looksLikeContentUri(sourcePath) ? sourcePath : '')
    || (looksLikeContentUri(storedPath) ? storedPath : '')
    || undefined
  const desktopPath = text(existing?.desktopPath)
    || (androidUri || looksLikeContentUri(sourcePath) ? '' : sourcePath)
  const displayPath = text(existing?.displayPath) || desktopPath || text(raw.title) || androidUri || storedPath
  const type = text(existing?.type) === 'android' || (Boolean(androidUri) && !desktopPath) ? 'android'
    : text(existing?.type) === 'display' ? 'display'
    : 'desktop'
  const locator: FileLocator = { type, displayPath }
  if (desktopPath) locator.desktopPath = desktopPath
  if (androidUri) locator.androidUri = androidUri
  return locator
}

export function normalizeFileMeta(
  raw: Record<string, unknown> | null | undefined,
  options: { sourceId?: string | null; now?: string } = {},
): FileMeta {
  const input = raw && typeof raw === 'object' ? raw : {}
  const now = options.now || new Date().toISOString()
  const id = text(input.id)
  const mode: FileMode = input.mode === 'copy' ? 'copy' : 'link'
  const kind: FileKind = input.kind === 'directory' ? 'directory' : 'file'
  const locatorInput = input.locator && typeof input.locator === 'object'
    ? input.locator as Record<string, unknown>
    : null
  const sourcePath = text(input.sourcePath)
    || text(locatorInput?.desktopPath)
    || text(locatorInput?.androidUri)
  const storedPath = text(input.storedPath) || (mode === 'link' ? sourcePath : '')
  const sourceId = text(input.sourceId) || text(options.sourceId) || undefined
  const hash = text(input.contentHash) || text(input.sha256) || null
  const meta: FileMeta = {
    id,
    title: text(input.title) || id,
    kind,
    mode,
    mime: text(input.mime) || 'application/octet-stream',
    ext: text(input.ext),
    size: Number(input.size || 0) || 0,
    sourcePath,
    storedPath,
    locator: locatorFrom(input, sourcePath, storedPath),
    contentHash: hash,
    sha256: hash,
    createdAt: text(input.createdAt) || now,
    updatedAt: text(input.updatedAt) || now,
  }
  if (sourceId) meta.sourceId = sourceId
  if (input.preview != null) meta.preview = typeof input.preview === 'string' ? input.preview : null
  if (input.entryCount != null) meta.entryCount = input.entryCount as FileMeta['entryCount']
  if (input.mtime != null) meta.mtime = text(input.mtime) || null
  return meta
}

export function fileMetaToJson(meta: FileMeta): Record<string, unknown> {
  const json: Record<string, unknown> = {
    id: meta.id,
    title: meta.title,
    kind: meta.kind,
    mode: meta.mode,
    mime: meta.mime,
    ext: meta.ext,
    size: meta.size,
    sourcePath: meta.sourcePath,
    storedPath: meta.storedPath,
    locator: meta.locator,
    createdAt: meta.createdAt,
    updatedAt: meta.updatedAt,
  }
  if (meta.sourceId) json.sourceId = meta.sourceId
  if (meta.contentHash) json.contentHash = meta.contentHash
  if (meta.sha256) json.sha256 = meta.sha256
  if (meta.preview != null) json.preview = meta.preview
  if (meta.entryCount != null) json.entryCount = meta.entryCount
  if (meta.mtime != null) json.mtime = meta.mtime
  return json
}

File: src-tauri/crates/tie-common/src/file_meta.rs
use serde_json::{json, Map, Value};

/// Fill `sourceId` / `locator` while keeping legacy `sourcePath` / `storedPath`.
pub fn normalize_file_meta(raw: &Value, source_id: Option<&str>) -> Value {
    let mut meta = match raw {
        Value::Object(map) => Value::Object(map.clone()),
        _ => json!({}),
    };
    let source_path = string_field(&meta, "sourcePath")
        .or_else(|| locator_string(&meta, "desktopPath"))
        .unwrap_or_default();
    let stored_path = string_field(&meta, "storedPath").unwrap_or_else(|| {
        if string_field(&meta, "mode").as_deref() != Some("copy") {
            source_path.clone()
        } else {
            String::new()
        }
    });
    let display_path = locator_string(&meta, "displayPath")
        .filter(|value| !value.is_empty())
        .unwrap_or_else(|| {
            if source_path.is_empty() {
                stored_path.clone()
            } else {
                source_path.clone()
            }
        });
    if meta.get("sourcePath").and_then(Value::as_str).unwrap_or("").is_empty() && !source_path.is_empty() {
        meta["sourcePath"] = json!(source_path);
    }
    if meta.get("storedPath").and_then(Value::as_str).unwrap_or("").is_empty() && !stored_path.is_empty() {
        meta["storedPath"] = json!(stored_path);
    }
    let resolved_source = string_field(&meta, "sourceId")
        .filter(|value| !value.is_empty())
        .or_else(|| source_id.map(str::trim).filter(|value| !value.is_empty()).map(str::to_string));
    if let Some(id) = resolved_source {
        meta["sourceId"] = json!(id);
    }
    if meta.get("locator").is_none() {
        let mut locator = Map::new();
        let is_android = source_path.starts_with("content://");
        locator.insert("type".into(), json!(if is_android { "android" } else { "desktop" }));
        if is_android {
            locator.insert("androidUri".into(), json!(source_path));
        } else if !source_path.is_empty() {
            locator.insert("desktopPath".into(), json!(source_path));
        }
        locator.insert("displayPath".into(), json!(display_path));
        meta["locator"] = Value::Object(locator);
    }
    if meta.get("contentHash").is_none() {
        if let Some(hash) = string_field(&meta, "sha256") {
            meta["contentHash"] = json!(hash);
        }
    }
    meta
}

pub fn source_path_from_meta(meta: &Value) -> String {
    string_field(meta, "sourcePath")
        .or_else(|| locator_string(meta, "desktopPath"))
        .or_else(|| locator_string(meta, "androidUri"))
        .unwrap_or_default()
}

fn string_field(meta: &Value, key: &str) -> Option<String> {
    meta.get(key)
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string)
}

fn locator_string(meta: &Value, key: &str) -> Option<String> {
    meta.get("locator")
        .and_then(Value::as_object)
        .and_then(|locator| locator.get(key))
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string)
}
